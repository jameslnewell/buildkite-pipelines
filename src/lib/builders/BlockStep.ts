import {BlockStepSchema, Field, StepDependsOn} from '../schema';
import {StepBuilder} from './StepBuilder';
import {BranchesBuilder, BranchesHelper} from './helpers/branches';
import {DependenciesBuilder, DependenciesHelper} from './helpers/dependencies';
import {KeyBuilder, KeyHelper} from './helpers/key';
import {LabelBuilder} from './helpers/label';
import {PromptBuilder, PromptHelper} from './helpers/prompt';

export class BlockStep
  implements
    StepBuilder,
    LabelBuilder,
    KeyBuilder,
    BranchesBuilder,
    DependenciesBuilder,
    PromptBuilder
{
  #label?: string;
  #state?: 'passed' | 'failed' | 'running';
  #keyHelper = new KeyHelper();
  #branchesHelper = new BranchesHelper();
  #dependenciesHelper = new DependenciesHelper();
  #promptHelper = new PromptHelper();

  label(label: string): this {
    this.#label = label;
    return this;
  }

  state(state: 'passed' | 'failed' | 'running'): this {
    this.#state = state;
    return this;
  }

  key(key: string): this {
    this.#keyHelper.key(key);
    return this;
  }

  branch(branch: string): this {
    this.#branchesHelper.branch(branch);
    return this;
  }

  dependOn(dependency: null | StepDependsOn): this {
    this.#dependenciesHelper.dependOn(dependency);
    return this;
  }

  allowDependencyFailure(allow: boolean): this {
    this.#dependenciesHelper.allowDependencyFailure(allow);
    return this;
  }

  prompt(prompt: string): this {
    this.#promptHelper.prompt(prompt);
    return this;
  }

  field(field: Field): this {
    this.#promptHelper.field(field);
    return this;
  }

  build(): BlockStepSchema | Promise<BlockStepSchema> {
    if (!this.#label) {
      throw new Error('BlockStep must have a label.');
    }

    const object: BlockStepSchema = {
      block: this.#label,
      ...this.#keyHelper.build(),
      ...this.#branchesHelper.build(),
      ...this.#dependenciesHelper.build(),
      ...this.#promptHelper.build(),
    };

    if (this.#state) {
      object.blocked_state = this.#state;
    }

    return object;
  }
}
