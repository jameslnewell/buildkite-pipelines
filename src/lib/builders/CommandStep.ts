import {StepBuilder} from './StepBuilder';
import {KeyBuilder, KeyHelper} from './helpers/key';
import {LabelBuilder, LabelHelper} from './helpers/label';
import {ConditionBuilder, ConditionHelper} from './helpers/condition';
import {BranchFilterBuilder, BranchFilterHelper} from './helpers/branches';
import {DependenciesBuilder, DependenciesHelper} from './helpers/dependencies';
import {SkipBuilder, SkipHelper} from './helpers/skip';
import {PluginBuilder} from './PluginBuilder';
import {CommandStepSchema, PluginSchema, StepDependsOn} from '../schema';
import {isPluginBuilder} from './isPluginBuilder';
import {AgentsBuilder, AgentsHelper} from './helpers/agents';
import {EnvironmentBuilder, EnvironmentHelper} from './helpers/env';
import {AutomaticRetry} from '../schema/schema';

type ManualRetry = Exclude<CommandStepSchema['retry'], undefined>['manual'];

export class CommandStep
  implements
    StepBuilder,
    KeyBuilder,
    LabelBuilder,
    ConditionBuilder,
    BranchFilterBuilder,
    DependenciesBuilder,
    SkipBuilder,
    AgentsBuilder,
    EnvironmentBuilder
{
  #commands?: string[];
  #plugins: Array<PluginSchema | PluginBuilder> = [];
  #keyHelper = new KeyHelper();
  #labelHelper = new LabelHelper();
  #conditionHelper = new ConditionHelper();
  #branchesHelper = new BranchFilterHelper();
  #dependenciesHelper = new DependenciesHelper();
  #skipHelper = new SkipHelper();
  #agentsHelper = new AgentsHelper();
  #envHelper = new EnvironmentHelper();

  #concurrency?: number;
  #concurrencyGroup?: string;
  #parallelism?: number;
  #softFail?: boolean;
  #timeoutInMinutes?: number;

  #automaticRetries: AutomaticRetry[] = [];
  #manualRetry: ManualRetry;

  /**
   * @deprecated Use .addCommand() instead
   */
  command(command: string): this {
    return this.addCommand(command);
  }

  addCommand(command: string): this {
    if (!this.#commands) {
      this.#commands = [];
    }
    this.#commands.push(command);
    return this;
  }

  /**
   * @deprecated Use .setKey() instead
   */
  key(key: string): this {
    return this.setKey(key);
  }

  setKey(key: string | undefined): this {
    this.#keyHelper.setKey(key);
    return this;
  }

  /**
   * @deprecated Use .setLabel() instead
   */
  label(label: string): this {
    return this.setLabel(label);
  }

  setLabel(label: string | undefined): this {
    this.#labelHelper.setLabel(label);
    return this;
  }

  /**
   * @deprecated Use .setCondition() instead
   */
  condition(condition: string): this {
    return this.setCondition(condition);
  }

  setCondition(condition: string | undefined): this {
    this.#conditionHelper.setCondition(condition);
    return this;
  }

  /**
   * @deprecated Use .addBranch() instead
   */
  branch(branch: string): this {
    return this.addBranch(branch);
  }

  addBranch(branch: string): this {
    this.#branchesHelper.addBranch(branch);
    return this;
  }

  /**
   * @deprecated Use .addDependency() instead
   */
  dependOn(dependency: null | StepDependsOn): this {
    return this.addDependency(dependency);
  }

  addDependency(dependency: null | StepDependsOn): this {
    this.#dependenciesHelper.addDependency(dependency);
    return this;
  }

  /**
   * @deprecated Use .addDependency() instead
   */
  allowDependencyFailure(allow: boolean): this {
    return this.setAllowDependencyFailure(allow);
  }

  setAllowDependencyFailure(allow: boolean | undefined): this {
    this.#dependenciesHelper.setAllowDependencyFailure(allow);
    return this;
  }

  /**
   * @deprecated Use .setSkip() instead
   */
  skip(skip: boolean | string): this {
    return this.setSkip(skip);
  }

  setSkip(skip: boolean | string | undefined): this {
    this.#skipHelper.setSkip(skip);
    return this;
  }

  /**
   * @deprecated Use .addPlugin() instead
   */
  plugin(plugin: PluginSchema | PluginBuilder): this {
    return this.addPlugin(plugin);
  }

  addPlugin(plugin: PluginSchema | PluginBuilder): this {
    this.#plugins.push(plugin);
    return this;
  }

  /**
   * @deprecated Use .addPlugins() instead
   */
  plugins(plugins: Iterable<PluginSchema | PluginBuilder>): this {
    return this.addPlugins(plugins);
  }

  addPlugins(plugins: Iterable<PluginSchema | PluginBuilder>): this {
    this.#plugins.push(...plugins);
    return this;
  }

  /**
   * @deprecated Use .setParallelism() instead
   */
  parallelism(parallelism: number): this {
    return this.setParallelism(parallelism);
  }

  setParallelism(parallelism: number | undefined): this {
    if (parallelism === 0) {
      throw new Error('Parallelism of zero will result in step being omitted');
    }

    this.#parallelism = parallelism;
    return this;
  }

  /**
   * @deprecated Use .addEnv() instead
   */
  env(name: string, value: unknown): this {
    return this.addEnv(name, value);
  }

  addEnv(name: string, value: unknown): this {
    this.#envHelper.addEnv(name, value);
    return this;
  }

  /**
   * @deprecated Use .setConcurrency() instead
   */
  concurrency(group: string, jobs: number): this {
    return this.setConcurrency(group, jobs);
  }

  setConcurrency(group: string, jobs: number): this {
    if (jobs === 0) {
      throw new Error(
        'Concurrency of zero will result in step which never starts',
      );
    }

    this.#concurrency = jobs;
    this.#concurrencyGroup = group;
    return this;
  }

  /**
   * @deprecated Use .setSoftFail() instead
   */
  softFail(fail: boolean = true): this {
    return this.setSoftFail(fail);
  }

  setSoftFail(fail: boolean | undefined): this {
    this.#softFail = fail;
    return this;
  }

  /**
   * @deprecated Use .setTimeout() instead
   */
  timeout(minutes: number): this {
    return this.setTimeout(minutes);
  }

  setTimeout(minutes: number | undefined): this {
    this.#timeoutInMinutes = minutes;
    return this;
  }

  /**
   * @deprecated Use .addAgent() instead
   */
  agent(tag: string, value: string): this {
    return this.addAgent(tag, value);
  }

  addAgent(tag: string, value: string): this {
    this.#agentsHelper.addAgent(tag, value);
    return this;
  }

  setManualRetry(retry: ManualRetry): this {
    this.#manualRetry = retry;
    return this;
  }

  addAutomaticRetry(retry: AutomaticRetry): this {
    this.#automaticRetries.push(retry);
    return this;
  }

  async build(): Promise<CommandStepSchema> {
    const object: CommandStepSchema = {
      ...{commands: this.#commands},
      ...this.#keyHelper.build(),
      ...this.#labelHelper.build(),
      ...this.#conditionHelper.build(),
      ...this.#branchesHelper.build(),
      ...this.#dependenciesHelper.build(),
      ...this.#skipHelper.build(),
      ...(this.#parallelism ? {parallelism: this.#parallelism} : {}),
      ...(this.#concurrency ? {concurrency: this.#concurrency} : {}),
      ...(this.#concurrencyGroup
        ? {concurrency_group: this.#concurrencyGroup}
        : {}),
      ...this.#envHelper.build(),
      ...(this.#softFail ? {soft_fail: this.#softFail} : {}),
      ...(this.#timeoutInMinutes
        ? {timeout_in_minutes: this.#timeoutInMinutes}
        : {}),
      ...this.#agentsHelper.build(),
    };

    if (this.#manualRetry !== undefined) {
      object.retry = {
        ...object.retry,
        manual: this.#manualRetry,
      };
    }
    if (this.#automaticRetries.length) {
      object.retry = {
        ...object.retry,
        automatic: this.#automaticRetries,
      };
    }

    if (this.#plugins.length > 0) {
      object.plugins = await Promise.all(
        this.#plugins.map(async (plugin) =>
          isPluginBuilder(plugin) ? await plugin.build() : plugin,
        ),
      );
    }

    return object;
  }
}
