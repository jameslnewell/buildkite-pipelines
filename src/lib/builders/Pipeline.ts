import {PipelineNotification, PipelineSchema, StepSchema} from '../schema';
import {PipelineBuilder} from './PipelineBuilder';
import {StepBuilder} from './StepBuilder';
import {StepsBuilder, StepsHelper} from './helpers/steps';
import {AgentsBuilder, AgentsHelper} from './helpers/agents';
import {NotifyBuilder, NotificationHelper} from './helpers/notifification';
import {EnvironmentBuilder, EnvironmentHelper} from './helpers/env';

export class Pipeline
  implements
    PipelineBuilder,
    AgentsBuilder,
    NotifyBuilder,
    StepsBuilder,
    EnvironmentBuilder
{
  #agentsHelper = new AgentsHelper();
  #envHelper = new EnvironmentHelper();
  #notifyHelper = new NotificationHelper();
  #stepsHelper = new StepsHelper();

  /**
   * @deprecated Use .addAgent() instead
   */
  agent(tag: string, value: string): this {
    return this.addAgent(tag, value);
  }

  /**
   * @see https://buildkite.com/docs/agent/v3/cli-start#agent-targeting
   */
  addAgent(tag: string, value: string): this {
    this.#agentsHelper.addAgent(tag, value);
    return this;
  }

  addEnv(name: string, value: unknown): this {
    this.#envHelper.addEnv(name, value);
    return this;
  }

  /**
   * @deprecated Use .addNotification() instead
   */
  notify(notification: PipelineNotification): this {
    return this.addNotification(notification);
  }

  /**
   * @see https://buildkite.com/docs/pipelines/notifications
   */
  addNotification(notification: PipelineNotification): this {
    this.#notifyHelper.addNotification(notification);
    return this;
  }

  /**
   * @deprecated Use .addStep() instead
   */
  step(step: StepSchema | StepBuilder): this {
    this.addStep(step);
    return this;
  }

  addStep(step: StepSchema | StepBuilder): this {
    this.#stepsHelper.addStep(step);
    return this;
  }

  /**
   * @deprecated Use .addSteps() instead
   */
  steps(steps: Iterable<StepSchema | StepBuilder>): this {
    this.addSteps(steps);
    return this;
  }

  addSteps(steps: Iterable<StepSchema | StepBuilder>): this {
    this.#stepsHelper.addSteps(steps);
    return this;
  }

  async build(): Promise<PipelineSchema> {
    const pipeline: PipelineSchema = {
      ...this.#agentsHelper.build(),
      ...this.#notifyHelper.build(),
      ...(await this.#stepsHelper.build()),
      ...this.#envHelper.build(),
    };
    return pipeline;
  }
}
