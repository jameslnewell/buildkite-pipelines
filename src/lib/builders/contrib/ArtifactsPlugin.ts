import {PluginSchema} from '../../schema';
import {PluginBuilder} from '../PluginBuilder';

/**
 * @see https://github.com/buildkite-plugins/artifacts-buildkite-plugin
 */
export class ArtifactsPlugin implements PluginBuilder {
  static PLUGIN = 'artifacts#v1.9.2';

  #downloads: string[] = [];
  #uploads: string[] = [];
  #skipOnStatus: number[] = [];

  #options: {[name: string]: unknown} = {};

  /**
   * @deprecated Use .setCompressed() instead
   */
  compressed(file: string): this {
    return this.setCompressed(file);
  }

  setCompressed(file: string): this {
    this.#options['compressed'] = file;
    return this;
  }

  /**
   * @deprecated Use .setDownload() instead
   */
  download(file: string): this {
    return this.setDownload(file);
  }

  setDownload(glob: string): this {
    this.#downloads.push(glob);
    return this;
  }

  /**
   * @deprecated Use .setUpload() instead
   */
  upload(glob: string): this {
    return this.setUpload(glob);
  }

  setUpload(glob: string): this {
    this.#uploads.push(glob);
    return this;
  }

  /**
   * @deprecated Use .setStep() instead
   */
  step(id: string): this {
    return this.setStep(id);
  }

  setStep(id: string): this {
    this.#options['step'] = id;
    return this;
  }

  /**
   * @deprecated Use .setIgnoreMissing() instead
   */
  ignoreMissing(ignore: boolean = true): this {
    return this.setIgnoreMissing(ignore);
  }

  setIgnoreMissing(ignore: boolean): this {
    this.#options['ignore-missing'] = ignore;
    return this;
  }

  /**
   * @deprecated Use .setSkipOnStatus() instead
   */
  skipOnStatus(status: number): this {
    return this.setSkipOnStatus(status);
  }

  setSkipOnStatus(status: number): this {
    this.#skipOnStatus.push(status);
    return this;
  }

  build(): PluginSchema | Promise<PluginSchema> {
    if (this.#downloads.length) {
      this.#options['download'] = this.#downloads;
    }

    if (this.#uploads.length) {
      this.#options['upload'] = this.#uploads;
    }

    if (this.#skipOnStatus.length) {
      this.#options['skip-on-status'] = this.#skipOnStatus;
    }

    return {
      [ArtifactsPlugin.PLUGIN]: this.#options,
    };
  }
}
