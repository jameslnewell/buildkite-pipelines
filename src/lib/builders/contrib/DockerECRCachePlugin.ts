import {PluginSchema} from '../../schema';
import {PluginBuilder} from '../PluginBuilder';

/**
 * @see https://github.com/seek-oss/docker-ecr-cache-buildkite-plugin
 */
export class DockerECRCachePlugin implements PluginBuilder {
  static PLUGIN = 'seek-oss/docker-ecr-cache#v2.2.0';

  #cacheOn: string[] = [];
  #secrets: string[] = [];
  #buildArgs: string[] = [];
  #options: {[name: string]: unknown} = {};

  /**
   * @deprecated Use .addCacheOn() instead
   */
  cacheOn(glob: string): this {
    return this.addCacheOn(glob);
  }

  addCacheOn(glob: string): this {
    this.#cacheOn.push(glob);
    return this;
  }

  /**
   * @deprecated Use .setDockerfile() instead
   */
  dockerfile(file: string): this {
    return this.setDockerfile(file);
  }

  setDockerfile(file: string): this {
    this.#options['dockerfile'] = file;
    return this;
  }

  /**
   * @deprecated Use .setTarget() instead
   */
  target(target: string): this {
    return this.setTarget(target);
  }

  setTarget(target: string): this {
    this.#options['target'] = target;
    return this;
  }

  /**
   * @deprecated Use .setSecret() instead
   */
  secret(secret: string): this {
    return this.setSecret(secret);
  }

  setSecret(secret: string): this {
    this.#secrets.push(secret);
    return this;
  }

  /**
   * @deprecated Use .addBuildArg() instead
   */
  buildArg(arg: string): this {
    return this.addBuildArg(arg);
  }

  addBuildArg(arg: string): this {
    this.#buildArgs.push(arg);
    return this;
  }

  /**
   * @deprecated Use .setMaxAgeDays() instead
   */
  maxAgeDay(days: number): this {
    return this.setMaxAgeDays(days);
  }

  setMaxAgeDays(days: number): this {
    this.#options['max-age-days'] = days;
    return this;
  }

  /**
   * @deprecated Use .setRegion() instead
   */
  region(region: string): this {
    return this.setRegion(region);
  }

  setRegion(region: string): this {
    this.#options['region'] = region;
    return this;
  }

  /**
   * @deprecated Use .setECRName() instead
   */
  ecrName(name: string): this {
    return this.setECRName(name);
  }

  setECRName(name: string): this {
    this.#options['ecr-name'] = name;
    return this;
  }

  build(): PluginSchema | Promise<PluginSchema> {
    if (this.#cacheOn.length) {
      this.#options['cache-on'] = this.#cacheOn;
    }

    if (this.#secrets.length) {
      this.#options['secrets'] = this.#secrets;
    }

    if (this.#buildArgs.length) {
      this.#options['build-args'] = this.#buildArgs;
    }

    return {
      [DockerECRCachePlugin.PLUGIN]: this.#options,
    };
  }
}
