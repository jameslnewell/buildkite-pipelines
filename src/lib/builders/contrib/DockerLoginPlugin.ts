import {PluginBuilder} from '../PluginBuilder';

export interface DockerLoginPluginSchema {
  server?: string;
  username: string;
  'password-env': string;
  retries?: number;
}

export class DockerLoginPlugin implements PluginBuilder {
  #username?: string;
  #password?: string;
  #server?: string;
  #retries?: number;

  username(username: string): this {
    this.#username = username;
    return this;
  }
  passwordEnv(passwordEnv: string): this {
    this.#password = passwordEnv;
    return this;
  }
  server(server: string): this {
    this.#server = server;
    return this;
  }
  retries(retries: number): this {
    this.#retries = retries;
    return this;
  }

  build() {
    return {
      'docker-login#v2.1.0': {
        username: this.#username,
        'password-env': this.#password,
        server: this.#server,
        retries: this.#retries,
      },
    };
  }
}
