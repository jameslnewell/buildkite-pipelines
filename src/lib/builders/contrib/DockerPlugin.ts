import {PluginSchema} from '../../schema';
import {PluginBuilder} from '../PluginBuilder';

/**
 * @see https://github.com/buildkite-plugins/docker-buildkite-plugin
 */
export class DockerPlugin implements PluginBuilder {
  static PLUGIN = 'docker#v5.11.0';

  #options: {[name: string]: unknown} = {};
  #commands: string[] = [];
  #envvars: string[] = [];
  #volumes: string[] = [];

  getImage(): string | undefined {
    return this.#options['image'] as string | undefined;
  }

  /**
   * @deprecated Use .setImage() instead
   */
  image(image: string) {
    return this.setImage(image);
  }

  setImage(image: string) {
    this.#options['image'] = image;
    return this;
  }

  getAlwaysPull(): boolean | undefined {
    return this.#options['always-pull'] as boolean | undefined;
  }

  /**
   * @deprecated Use .setAlwaysPull() instead
   */
  alwaysPull(alwaysPull: boolean = true): this {
    return this.setAlwaysPull(alwaysPull);
  }

  setAlwaysPull(alwaysPull: boolean): this {
    this.#options['always-pull'] = alwaysPull;
    return this;
  }

  getCommands(): ReadonlyArray<string> {
    return this.#commands;
  }

  /**
   * @deprecated Use .addCommand() instead
   */
  command(command: string): this {
    return this.addCommand(command);
  }

  addCommand(command: string): this {
    this.#commands.push(command);
    return this;
  }

  getDebug(): boolean | undefined {
    return this.#options['debug'] as boolean | undefined;
  }

  /**
   * @deprecated Use .setDebug() instead
   */
  debug(debug: boolean = true): this {
    return this.setDebug(debug);
  }

  setDebug(debug: boolean): this {
    this.#options['debug'] = debug;
    return this;
  }

  getEntrypoint(): string | undefined {
    return this.#options['entrypoint'] as string | undefined;
  }

  /**
   * @deprecated Use .addCommand() instead
   */
  entrypoint(entrypoint: string): this {
    return this.setEntrypoint(entrypoint);
  }

  setEntrypoint(entrypoint: string): this {
    this.#options['entrypoint'] = entrypoint;
    return this;
  }

  getEnvironment(): ReadonlyArray<string> {
    return this.#envvars;
  }

  /**
   * @deprecated Use .addEnvironment() instead
   */
  environment(environment: string): this {
    return this.addEnvironment(environment);
  }

  addEnvironment(environment: string): this {
    this.#envvars.push(environment);
    return this;
  }

  getPropagateEnvironment(): boolean | undefined {
    return this.#options['propagate-environment'] as boolean | undefined;
  }

  /**
   * @deprecated Use .setPropagateEnvironment() instead
   */
  propagateEnvironment(propagate: boolean = true): this {
    return this.setPropagateEnvironment(propagate);
  }

  setPropagateEnvironment(propagate: boolean): this {
    this.#options['propagate-environment'] = propagate;
    return this;
  }

  getPropagateAWSAauthTokens(): boolean | undefined {
    return this.#options['propagate-aws-auth-tokens'] as boolean | undefined;
  }

  /**
   * @deprecated Use .setPropagateAWSAuthTokens() instead
   */
  propagateAWSAuthTokens(propagate: boolean = true): this {
    return this.setPropagateAWSAuthTokens(propagate);
  }

  /**
   * @deprecated Use .setPropagateAWSAuthTokens() instead
   */
  setPropagateAWSAauthTokens(propagate: boolean): this {
    return this.setPropagateAWSAuthTokens(propagate);
  }

  setPropagateAWSAuthTokens(propagate: boolean): this {
    this.#options['propagate-aws-auth-tokens'] = propagate;
    return this;
  }

  getMountCheckout(): boolean | undefined {
    return this.#options['mount-checkout'] as boolean | undefined;
  }

  /**
   * @deprecated Use .setMountCheckout() instead
   */
  mountCheckout(mount: boolean = true): this {
    return this.setMountCheckout(mount);
  }

  setMountCheckout(mount: boolean = true): this {
    this.#options['mount-checkout'] = mount;
    return this;
  }

  getMountBuildkiteAgent(): boolean | undefined {
    return this.#options['mount-buildkite-agent'] as boolean | undefined;
  }

  /**
   * @deprecated Use .setMountBuildkiteAgent() instead
   */
  mountBuildkiteAgent(mount: boolean = true): this {
    return this.setMountBuildkiteAgent(mount);
  }

  setMountBuildkiteAgent(mount: boolean): this {
    this.#options['mount-buildkite-agent'] = mount;
    return this;
  }

  getMountSSHAgent(): boolean | undefined {
    return this.#options['mount-ssh-agent'] as boolean | undefined;
  }

  /**
   * @deprecated Use .setMountSSHAgent() instead
   */
  mountSSHAgent(mount: boolean = true): this {
    return this.setMountSSHAgent(mount);
  }

  setMountSSHAgent(mount: boolean): this {
    this.#options['mount-ssh-agent'] = mount;
    return this;
  }

  getPlatform(): string | undefined {
    return this.#options['platform'] as string | undefined;
  }

  /**
   * @deprecated Use .setPlatform() instead
   */
  platform(platform: string): this {
    return this.setPlatform(platform);
  }

  setPlatform(platform: string): this {
    this.#options['platform'] = platform;
    return this;
  }

  getShell(): string[] | undefined {
    return this.#options['shell'] as string[] | undefined;
  }

  /**
   * @deprecated Use .setPlatform() instead
   */
  shell(cmd: string, args: string[] = []): this {
    return this.setShell(cmd, args);
  }

  setShell(cmd: string, args: string[] = []): this {
    this.#options['shell'] = [cmd, ...args];
    return this;
  }

  /**
   * @deprecated Use .setUser() instead
   */
  user(user: string): this {
    return this.setUser(user);
  }

  setUser(user: string): this {
    this.#options['user'] = user;
    return this;
  }

  /**
   * @deprecated Use .setUserNamespace() instead
   */
  userns(namespace: string): this {
    return this.setUserns(namespace);
  }

  setUserns(namespace: string): this {
    this.#options['userns'] = namespace;
    return this;
  }

  getVolumes(): ReadonlyArray<string> {
    return this.#volumes;
  }

  /**
   * @deprecated Use .addVolume() instead
   */
  volume(volume: string): this {
    return this.addVolume(volume);
  }

  addVolume(volume: string): this {
    this.#volumes.push(volume);
    return this;
  }

  /**
   * @deprecated Use .setWorkdir() instead
   */
  workdir(workdir: string): this {
    return this.setWorkdir(workdir);
  }

  setWorkdir(workdir: string): this {
    this.#options['workdir'] = workdir;
    return this;
  }

  /**
   * @deprecated Use .setCPUs() instead
   */
  cpus(cpus: string): this {
    return this.setCPUs(cpus);
  }

  setCPUs(cpus: string): this {
    this.#options['cpus'] = cpus;
    return this;
  }

  /**
   * @deprecated Use .setMemory() instead
   */
  memory(memory: string): this {
    return this.setMemory(memory);
  }

  setMemory(memory: string): this {
    this.#options['memory'] = memory;
    return this;
  }

  /**
   * @deprecated Use .setMemorySwap() instead
   */
  memorySwap(swap: string): this {
    return this.setMemorySwap(swap);
  }

  setMemorySwap(swap: string): this {
    this.#options['memory-swap'] = swap;
    return this;
  }

  /**
   * @deprecated Use .setMemorySwapiness() instead
   */
  memorySwapiness(swappiness: string): this {
    return this.setMemorySwapiness(swappiness);
  }

  setMemorySwapiness(swappiness: string): this {
    this.#options['memory-swapiness'] = swappiness;
    return this;
  }

  setExpandVolumeVars(expand: boolean): this {
    this.#options['expand-volume-vars'] = expand;
    return this;
  }

  build(): PluginSchema | Promise<PluginSchema> {
    const object: Record<string, unknown> = {
      ...this.#options,
    };

    if (this.#commands.length) {
      object['command'] = this.#commands;
    }

    if (this.#envvars.length) {
      object['environment'] = this.#envvars;
    }

    if (this.#volumes.length) {
      object['volumes'] = this.#volumes;
    }

    return {
      [DockerPlugin.PLUGIN]: object,
    };
  }
}
