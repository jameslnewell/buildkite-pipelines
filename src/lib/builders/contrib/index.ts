export * from './ECRPlugin';
export * from './DockerPlugin';
export * from './DockerLoginPlugin';
export * from './SecretsManagerPlugin';
export * from './DockerECRCachePlugin';
