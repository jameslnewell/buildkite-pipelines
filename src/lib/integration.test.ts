import {stringify, validate} from '.';
import {CommandStep, GroupStep, WaitStep} from './builders';
import {BlockStep} from './builders/BlockStep';
import {Pipeline} from './builders/Pipeline';
import {DockerPlugin} from './builders/contrib';

describe('integration', () => {
  test('matches snapshot', () => {
    const pipeline = new Pipeline()
      .step(
        new GroupStep()
          .label(':eslint: Lint group')
          .step(
            new CommandStep().label(':eslint: Lint').command('npm run lint'),
          ),
      )
      .step(
        new CommandStep()
          .label(':jest: Test')
          .command('npm run test')
          .key('unit-test'),
      )
      .step(
        new CommandStep()
          .label(':upload: Upload coverage')
          .agent('queue', 'arm')
          .command('npm run upload:coverage')
          .dependOn('unit-test')
          .plugin(new DockerPlugin().image('codeclimate/codeclimate')),
      )
      .step(new WaitStep())
      .step(new BlockStep().label('ðŸš€ Release').key('release'));
    const object = pipeline.build();
    expect(validate(object)).toHaveLength(0);
    expect(stringify(object)).toMatchSnapshot();
  });
});
