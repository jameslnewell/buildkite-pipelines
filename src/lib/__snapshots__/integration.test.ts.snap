// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration matches snapshot 1`] = `
"steps:
  - group: \\":eslint: Lint group\\"
    label: \\":eslint: Lint group\\"
    steps:
      - commands:
          - npm run lint
        label: \\":eslint: Lint\\"
  - commands:
      - npm run test
    key: unit-test
    label: \\":jest: Test\\"
  - commands:
      - npm run upload:coverage
    label: \\":upload: Upload coverage\\"
    depends_on:
      - unit-test
    agents:
      queue: arm
    plugins:
      - docker#v5.11.0:
          image: codeclimate/codeclimate
  - wait: null
  - block: ðŸš€ Release
    key: release
"
`;
